apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/RetryTest.cs

```diff
@@ -35,61 +35,95 @@ public class RetryTest
+            maximumRetries: 3,
+            success: success,
+            expectedBackOffs: new int[] { 0, 1, 3 });
     }
 
     [Theory]
     [InlineData(0, 1, 2, 0, 0)]
     [InlineData(1, 2, 2, 2, 0, 1, 3)]
     [InlineData(1, 4, 1, 3, 0, 1, 2, 3)]
-    public void CustomRetryTimingTest(int initialBackoff, int maxBackoff, double backoffMultiplier, int expectedTotalRetries, params int[] expectedBackOffs)
+    public void CustomRetryTimingTest(int initialBackoff, int maxBackoff, double backoffMultiplier, int numOfFailures, params int[] expectedBackOffs)
     {
         RetryOptions retryOptions = new RetryOptions(
             retryTiming: new RetryTiming(initialBackoff: TimeSpan.FromSeconds(initialBackoff),
             maxBackoff: TimeSpan.FromSeconds(maxBackoff), backoffMultiplier: backoffMultiplier),
             retryPredicate: RetryPredicate.FromErrorCodes(502, 504));
 
-        AssertAttempts(retryOptions, 502, expectedTotalRetries, true, expectedBackOffs);
+        AssertAttempts(
+            retryOptions: retryOptions,
+            statusCode: 502,
+            numOfFailures: numOfFailures,
+            maximumRetries: numOfFailures + 1,
+            success: true,
+            expectedBackOffs: expectedBackOffs);
     }
 
-    private static void AssertAttempts(RetryOptions retryOptions, int statusCode, int expectedTotalRetries, bool success, params int[] expectedBackOffs)
+    [Fact]
+    public void ExceptionAfterRetryExhaustedTest()
+    {
+        RetryOptions retryOptions = new RetryOptions(
+            retryTiming: new RetryTiming(initialBackoff: TimeSpan.FromSeconds(1),
+            maxBackoff: TimeSpan.FromSeconds(3), backoffMultiplier: 2),
+            retryPredicate: RetryPredicate.FromErrorCodes(502, 504));
+
+        var expectedBackOffs = new[] { 0, 1, 3 };
+
+        AssertAttempts(
+           retryOptions: retryOptions,
+           statusCode: 502,
+           numOfFailures: 4,
+           maximumRetries: 3,
+           success: false,
+           expectedBackOffs: expectedBackOffs);
+    }
+
+    private static void AssertAttempts(RetryOptions retryOptions, int statusCode, int numOfFailures, int maximumRetries, bool success, params int[] expectedBackOffs)
     {
         var messageHandler = new ReplayingMessageHandler(VersionHeaderBuilder.HeaderName);
         var service = new FakeStorageService(messageHandler);
         service.HttpClient.MessageHandler.GoogleApiClientHeader = "test/fake";
-        service.HttpClient.MessageHandler.NumTries = expectedTotalRetries + 1;
+        service.HttpClient.MessageHandler.NumTries = maximumRetries;
 
         var request = service.Buckets.Get("bucket");
         var client = new StorageClientImpl(service);
 
-        // Retries for the the failures.
-        service.ExpectRequests(request, (HttpStatusCode) statusCode, expectedTotalRetries);
+        // Retries for the the failures. Assumed that error code 504 is always included in the predicate for these tests.
```

Comment:
I would personally have expressed the test with an array of status codes to return, one per request. That way you can end with a success, or end with an error code that *isn't* matched by a predicate. It puts the flexibility in the hands of the caller, keeping this common method simpler.

Reply:
As discussed, we can take up improving the integration tests in a separate PR. Merging the library changes for now.